library/data/registry_dicts/src/main/java/org/quiltmc/qsl/registry/dict/impl/reloader/DictMap.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.registry.dict.impl.reloader;
+
+import static org.quiltmc.qsl.registry.dict.impl.reloader.RegistryDictReloader.LOGGER;
+
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.JsonOps;
+import org.apache.logging.log4j.Level;
+
+import net.minecraft.resource.Resource;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.InvalidIdentifierException;
+import net.minecraft.util.JsonHelper;
+import net.minecraft.util.registry.Registry;
+
+import org.quiltmc.qsl.registry.dict.api.RegistryDict;
+
+final class DictMap {
+	private final Registry<?> registry;
+	private final RegistryDict<?, ?> dictionary;
+	private final TagGetter tagGetter;
+	private final Map<DictTarget, Object> map;
+
+	public DictMap(Registry<?> registry, RegistryDict<?, ?> dictionary, TagGetter tagGetter) {
+		this.registry = registry;
+		this.dictionary = dictionary;
+		this.tagGetter = tagGetter;
+		map = new HashMap<>();
+	}
+
+	public void put(Identifier id, Object value) {
+		map.put(new DictTarget.Single(id), value);
+	}
+
+	public void putTag(Identifier tagId, Object value, boolean required) {
+		map.put(new DictTarget.Tagged<>(tagGetter, registry, tagId, required), value);
+	}
+
+	public Registry<?> getRegistry() {
+		return registry;
+	}
+
+	public RegistryDict<?, ?> getDictionary() {
+		return dictionary;
+	}
+
+	public Map<DictTarget, Object> getMap() {
+		return map;
+	}
+
+	// FIXME tag support!
```

Comment:
Does this need to be fixed?

Reply:
Oh, right. I... never actually finished this part. Well, what with tags basically being completely rewritten, might as well do it now!

Reply:
...no, wait, I _did_ do tag support! I just forgot to remove this.
But I'll probably end up rewriting tag support because of the 1.18.2 refactor anyway.

Reply:
AGH, forgot to remove this again